### ISSUE TITLE
Finamp consumes a lot of battery while pausing in the background

### ISSUE THREAD

**Issue by ha-ku** [jmshrv/finamp#795]
I used Finamp to play music for a while during the day and kept Finamp in the background. That night I went to sleep with 53% battery on the phone and woke up with only 4%. The system's battery usage stats showed that Finamp used close to 40% of that power in the background that night, but I didn't use it to play anything all night.
Finamp version: 0.9.7
Android version: 14
![Screenshot_20240623-112611](https://github.com/jmshrv/finamp/assets/18721230/737d2cbc-0e31-414d-ada4-eac2bf9bb980)

**Comment by tibequadorian** [comment]
Also happens on 0.9.8

**Comment by Chaphasilor** [comment]
That is probably because you have "Enter low-priority state on pause" disabled in the Audio Service settings (which is also the default). This is used to make sure Finamp is not killed in the background, so you can start playing even after having music paused for a few hours. I'm not sure why exactly it's draining so much battery in your case, but increased battery drain is definitely expected.  
If you can upload logs of the issue I could try to figure out what's causing the massive battery drain, but if you don't mind re-opening Finamp after a few minutes of inactivity, I'd suggest you simply turn on the setting!

**Comment by tibequadorian** [comment]
It does not happen on the current stable version (0.6.23)

**Comment by BBaoVanC** [comment]
It seems to me that enabling the "Enter low-priority state on pause" option has fixed the battery drain. I haven't ever had the issue again since turning it on. Is there a reason it isn't defaulted to on?

**Comment by Chaphasilor** [comment]
The reason is that it was changed before Finamp had the ability to restore the queue, so keeping the app alive was the only way to keep the queue.  
Also, it is only possible to resume music after a while via e.g. bluetooth headphones if the app is still active. Sometimes the app will get started after it was killed when you try to resume, but then you have to press resume a second time for it to actually start playing again.

But given the battery drain issues, I might change the default again :)

**Comment by Chaphasilor** [comment]
The default for this setting will be changed with the next beta update, back to how it is on the regular/stable version.  
For existing installations you'll have to **manually** turn on the setting (Settings > "Audio Service" > "Enter Low-Priority Mode on Pause")

**Comment by ravtakhar** [comment]
This is still happening on a Google Pixel 7 Pro, Samsung S22, Google Pixel 6, Google Pixel 3A and Google Pixel 3
All with Enter Low Priority Mode enabled

**Comment by Chaphasilor** [comment]
@ravtakhar without any playback? Does the app have any exemptions from battery optimizations?

**Comment by ravtakhar** [comment]
Yeah, this is happening without playback but another thing I am noticing is that when the app is closed (multi-tasking) it will actually not turn off music. Still required to pause the music from the notification tray.

This is only happening on the Google Pixel 3A and Google Pixel 3 which have the same battery optimization as the Google Pixel 7 Pro & Google Pixel 6.

**Comment by TheFeelTrain** [comment]
This is still a problem on the 0.9.19 beta

<img width="1080" height="237" alt="Image" src="https://github.com/user-attachments/assets/8312db56-e431-4ccd-a398-198f268c1467" />

**Comment by HeroBrine1st** [comment]
First of all, the battery drain is *estimation* of actual battery drain. Android likely has no means of actually measuring energy consumption by processes.

This proposes the first question: is this anomaly simply this high number? Or it *really* drains the battery?

Actually I was suprised I am subscribed to this issue. I checked and yes, I am affected, but I see no battery drain, system successfully goes into deep sleep and all the essentials. So I completely forgot about this issue, most likely I was subscribed because "Enter Low-Priority Mode on Pause" being off prevents deep sleep, but anyway.

I have some vital observations. I have debug application build on my phone (one test is pending so I didn't delete it to avoid relogin). According to android battery drain screen, it drained 2% while release github build (which I use daily) used 29%. Debug build has no player (and no queue to restore, I opened it to check), while release build was ready to play at my command, no "finamp is running" etc (and this didn't prevent deep sleep in the last 8 minutes (time from last charging)). I have "Enter Low-Priority Mode on Pause" enabled and android kills the finamp sometimes, so this feature works normally. 

So, the hypothesis: the player is [foreground service](https://developer.android.com/develop/background-work/services/fgs). It is likely paused, but not destroyed (notification is visible and service is ready to playback). Battery measurement counts that as battery consumption but service doesn't do anything.

Btw, as I have discovered, it is definitely not Riverpod because it disables itself in background.

---

So, do someone really feel the battery drain? Does the situation change (including real battery drain) if playback is stopped entirely (i.e. no queue at all as opposed to having it on pause)? Don't other players show there having high drain too? Does stopping those players entirely fix this issue with them?

**Comment by HeroBrine1st** [comment]
For future reports. I think this info would be relevant:

- For Finamp (open "App info" in android settings) you should have App battery usage -> Optimised (recommended for most apps)
- Is player notification visible? (if so, is it an Android 12 "pinned media player" with only "Play" button or fully working notification?)
- Does the phone go into deep sleep without problems? (I use AccuBattery to get that, something like 60% of screen off time given long standby is okay, but if it is finamp it will entirely prevent it)
- If you see low deep sleep and if you know how to see wakelocks on your device (starting with Android 15 it is somewhat impossible without system (user 1000, a.k.a. adb shell) privileges), does Finamp keep a wakelock?
  Actually, it is either you didn't enable "Enter Low-Priority Mode on Pause" or it keeps a wakelock, so this may be excessive
- Is your OS AOSP-like (can include Google Services) or it has heavy modifications? AOSP-like operating systems usually don't have unique names, it may be "Android One" or "Pure Android" (as opposed to MIUI, OxygenOS etc)
- Actual battery usage by the whole system (approximately substract percentages from status bar) and the time phone was in standby

If you don't use finamp actively, I recommend starting the test with disconnecting the charger (so that AccuBattery or other measurement utility starts measurement from known point of time), playing something and then leaving the phone on standby.

Copy this template:

```
- [ ] Finamp battery usage setting is set to "optimised" (recommended for most apps)
- [ ] Player notification is there and does not have anything except "Play" button
- [ ] Player notification is gone
- [ ] Phone sleeps deeply without problems
  - [ ] (optional, remove if unanswered) Finamp keeps a wakelock and so prevents deep sleeping
  - [ ] "Enter Low-Priority Mode on Pause" is enabled in Finamp settings
- Operating system: 
- Period of time phone was standby (wall clock time): 
- Approximately, how much charge was drained by whole system while the phone was standby: 
```

If you can't provide all of that but still want to report, remove relevant items except last two items (they are used to determine if you have the problem at all).

@Chaphasilor pinging you because i may be scaring affected users with this. I think you have better experience at that if I did anything wrong

**Comment by Chaphasilor** [comment]
@HeroBrine1st so let me give a brief overview:  

1. If Finamp is playing music, it runs as a foreground service
2. If Finamp is paused, then depending on the ["Enter Low-Priority State on Pause"](https://intradeus.github.io/http-protocol-redirector?r=finamp://internal/settings/audioservice) setting, Finamp will either go into the background (if setting is **on**), or stay in the foreground (if setting is **off**). So disabling this setting will keep Finamp active even if paused, so that music can always be resumed (unless the system actively kills the process). If the setting is on, then the system is free to kill Finamp at any time, even without "aggressive" background management
3. If "Enter Low-Priority State on Pause" is **off**, additional battery drain is expected. Android should report that as "Foreground". If "Enter Low-Priority State on Pause" is **on**, battery drain is *unexpected*, since the system is allowed to kill the app to save battery. Usage here should be reported as "Background"
4. Some of the reports (e.g. OP) explicitly mention the battery *actually draining more quickly than expected*, so it seems to be a real issue. And that 24h of background usage also seems strange.
5. Filling out your template would definitely help, if someone still has an issue
